relint.py

```diff
@@ -69,38 +131,45 @@ def main():
 
     tests = list(load_config(args.config))
 
+    diff = None
+    if args.diff:
+        git_command = ['git', 'diff', '-U0', '--cached']
+        output = subprocess.check_output(git_command, encoding='utf-8')
+        diff = parse_diff(output)
+
     matches = chain.from_iterable(
-        lint_file(path, tests)
+        lint_file(path, tests, diff)
         for path in paths
     )
-
     _filename = ''
     lines = []
 
     exit_code = 0
 
-    for filename, test, match in matches:
-        exit_code = test.error if exit_code == 0 else exit_code
-        if filename != _filename:
-            _filename = filename
-            lines = match.string.splitlines()
-
-        start_line_no = match.string[:match.start()].count('\n')
-        end_line_no = match.string[:match.end()].count('\n')
-        output_format = "{filename}:{line_no} {test.name}"
-        print(output_format.format(
-            filename=filename, line_no=start_line_no + 1, test=test,
-        ))
-        if test.hint:
-            print("Hint:", test.hint)
-        match_lines = (
-            "{line_no}>    {code_line}".format(
-                line_no=no + start_line_no + 1,
-                code_line=line,
+    if diff == {}:  # when nothing is cached
```

Comment:
I don't think that any of those changes should be needed. You should have less matches to begin with. The behavior for each match should be the same.